// @wdio/browserstack-service/build/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS = exports.DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS = exports.BATCH_EVENT_TYPES = exports.DATA_BATCH_INTERVAL = exports.DATA_BATCH_SIZE = exports.DATA_SCREENSHOT_ENDPOINT = exports.DATA_BATCH_ENDPOINT = exports.DATA_EVENT_ENDPOINT = exports.DATA_ENDPOINT = exports.DEFAULT_OPTIONS = exports.VALID_APP_EXTENSION = exports.BROWSER_DESCRIPTION = void 0;
exports.BROWSER_DESCRIPTION = [
    'device',
    'os',
    'osVersion',
    'os_version',
    'browserName',
    'browser',
    'browserVersion',
    'browser_version'
];
exports.VALID_APP_EXTENSION = [
    '.apk',
    '.aab',
    '.ipa'
];
exports.DEFAULT_OPTIONS = {
    setSessionName: true,
    setSessionStatus: true,
    testObservability: true
};
exports.DATA_ENDPOINT = 'https://collector-observability.browserstack.com';
exports.DATA_EVENT_ENDPOINT = 'api/v1/event';
exports.DATA_BATCH_ENDPOINT = 'api/v1/batch';
exports.DATA_SCREENSHOT_ENDPOINT = 'api/v1/screenshots';
exports.DATA_BATCH_SIZE = 1000;
exports.DATA_BATCH_INTERVAL = 2000;
exports.BATCH_EVENT_TYPES = ['LogCreated', 'TestRunStarted', 'TestRunFinished', 'HookRunFinished', 'HookRunStarted', 'ScreenshotCreated'];
exports.DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS = 5000; // 5s
exports.DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS = 100; // 100ms
